server.c

```diff
@@ -0,0 +1,282 @@
+{
+  if (send_wrapper (socket_fd, task, sizeof (task_t), 0) == S_FAILURE)
+    {
+      print_error ("Could not send data to client\n");
+      return (S_FAILURE);
+    }
+
+  print_error ("Sent task %s to client\n", task->password);
+
+  int size;
+  if (recv_wrapper (socket_fd, &size, sizeof (int), 0) == S_FAILURE)
+    {
+      print_error ("Could not receive data from client\n");
+      return (S_FAILURE);
+    }
+
+  print_error ("Received %d from client\n", size);
+
+  if (size != 0)
+    {
+      if (recv_wrapper (socket_fd, password, sizeof (password_t), 0)
+          == S_FAILURE)
+        {
+          print_error ("Could not receive data from client\n");
+          return (S_FAILURE);
+        }
+
+      print_error ("Received password %s from client\n", password);
+    }
+
+  return (S_SUCCESS);
+}
+
+static void *
+handle_client (void *arg)
+{
+  cl_context_t *cl_ctx = (cl_context_t *)arg;
+  cl_context_t local_ctx = *cl_ctx;
+  mt_context_t *mt_ctx = &local_ctx.context->context;
+
+  if (pthread_mutex_unlock (&cl_ctx->mutex) != 0)
+    {
+      print_error ("Could not unlock mutex\n");
+      return (NULL);
+    }
+
+  print_error ("Mutex unlocked\n");
+
+  if (send_wrapper (local_ctx.socket_fd, mt_ctx->config->hash, HASH_LENGTH, 0)
+      == S_FAILURE)
+    {
+      print_error ("Could not send hash to client\n");
+      return (NULL);
+    }
+
+  print_error ("Sent hash to client\n");
+
+  while (true)
+    {
+      task_t task;
+      if (queue_pop (&mt_ctx->queue, &task) == S_FAILURE)
+        return (NULL);
+
+      task.to = task.from;
+      task.from = 0;
+
+      if (delegate_task (local_ctx.socket_fd, &task, mt_ctx->password)
+          == S_FAILURE)
+        {
+          if (queue_push (&mt_ctx->queue, &task) == S_FAILURE)
+            print_error ("Could not push to the queue\n");
+          return (NULL);
+        }
+
+      if (pthread_mutex_lock (&cl_ctx->mutex) != 0)
+        {
+          print_error ("Could not lock a mutex\n");
+          return (NULL);
+        }
+      pthread_cleanup_push (cleanup_mutex_unlock, &cl_ctx->mutex);
+
+      --mt_ctx->passwords_remaining;
+      if (mt_ctx->passwords_remaining == 0 || mt_ctx->password[0] != 0)
+        if (pthread_cond_signal (&mt_ctx->cond_sem) != 0)
+          {
+            print_error ("Could not signal a condition\n");
+            return (NULL);
+          }
+
+      pthread_cleanup_pop (!0);
+    }
+
+  close (local_ctx.socket_fd);
+
+  return (NULL);
+}
+
+static void *
+handle_clients (void *arg)
+{
+  serv_context_t *serv_ctx = (serv_context_t *)arg;
+  mt_context_t *mt_ctx = &serv_ctx->context;
+
+  while (true)
+    {
+      int client_socket = accept (serv_ctx->socket_fd, NULL, NULL);
+      if (client_socket == -1)
+        {
+          print_error ("Could not accept new connection\n");
+          continue;
+        }
+
+      print_error ("Accepted new connection\n");
+
+      cl_context_t cl_ctx = {
```

Comment:
Move client context and mutex initialisation out of the loop. Don't forget to unlock mutex in case if thread failed to start.